predicators/train_refinement_estimator.py

```diff
@@ -0,0 +1,285 @@
+    refinement_estimator.train(dataset)
+    train_time = time.perf_counter() - train_start_time
+    logging.info(f"Finished training in {train_time:.5f} seconds")
+    # Save the training state to a file
+    # Create saved data directory.
+    os.makedirs(CFG.approach_dir, exist_ok=True)
+    config_path_str = utils.get_config_path_str()
+    state_file = f"{CFG.refinement_estimator}_{config_path_str}.estimator"
+    state_file_path = Path(CFG.approach_dir) / state_file
+    refinement_estimator.save_state(state_file_path)
+    logging.info(f"Saved trained estimator to {state_file_path}")
+
+    script_time = time.perf_counter() - script_start
+    logging.info(f"\n\nScript terminated in {script_time:.5f} seconds")
+
+
+def _generate_refinement_data(
+        env: BaseEnv, preds: Set[Predicate], options: Set[ParameterizedOption],
+        train_tasks: List[Task]) -> List[RefinementDatapoint]:
+    """Collect refinement data and save the dataset to a file."""
+    num_tasks = len(train_tasks)
+    nsrts = get_gt_nsrts(CFG.env, preds, options)
+    option_model = create_option_model(CFG.option_model_name)
+
+    # Generate the dataset and save it to file.
+    dataset: List[RefinementDatapoint] = []
+    for test_task_idx, task in enumerate(train_tasks):
+        try:
+            _collect_refinement_data_for_task(task, option_model, nsrts, preds,
+                                              env.types,
+                                              CFG.seed + test_task_idx,
+                                              dataset)
+            logging.info(f"Task {test_task_idx+1} / {num_tasks}: Success")
+        except (PlanningTimeout, _SkeletonSearchTimeout) as e:
+            logging.info(f"Task {test_task_idx+1} / {num_tasks} failed by "
+                         f"timing out: {e}")
+    logging.info(f"Got {len(dataset)} data points.")
+    # Create saved data directory.
+    os.makedirs(CFG.data_dir, exist_ok=True)
+    # Create file path.
+    data_file_path = _get_data_file_path()
+    # Store the train tasks just in case we need it in the future.
+    # (Note: unpickling this doesn't work...)
+    # data_content = {
+    #     "tasks": train_tasks,
+    #     "data": dataset,
+    # }
+    logging.info(f"Writing dataset to {data_file_path}")
+    with open(data_file_path, "wb") as f:
+        pkl.dump(dataset, f)
+    return dataset
+
+
+def _collect_refinement_data_for_task(task: Task,
+                                      option_model: _OptionModelBase,
+                                      nsrts: Set[NSRT],
+                                      predicates: Set[Predicate],
+                                      types: Set[Type], seed: int,
+                                      data: List[RefinementDatapoint]) -> None:
+    """Generate refinement data from given train task and append to data.
+
+    Adapted from _sesame_plan_with_astar() from predicators/planning.py.
+    """
+    ground_nsrt_timeout = CFG.timeout
```

Comment:
this is quite a bit of copied code, so let's think about whether there's another way. if we're willing to make changes within `planning.py`, could we log the metrics we need inside of `metrics`?

Reply:
my brain is out of thinking juice for the week but I can return to this later if there isn't a clear answer